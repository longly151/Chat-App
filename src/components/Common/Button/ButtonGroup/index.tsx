import React from 'react';
import QuickView from '../../View/QuickView';
import Button, { ButtonProps } from '../DefaultButton';
import FlatList, { FlatListProps } from '../../FlatList/DefaultFlatList';

export interface ButtonGroupProps extends Omit<ButtonProps, 'title'|'t'|'onPress'|'margin'|'marginLeft'|'marginVertical'>{
  defaultIndexChange?: boolean;
  titleList?: Array<string>;
  onItemPress?: (index: number) => any;
  flatListProps?: FlatListProps
  defaultActiveIndex?: number;
}

interface State {
  activeIndex: number;
}

class ButtonGroup extends React.Component<ButtonGroupProps, State> {
  static defaultProps = {
    defaultIndexChange: true,
  };

  private flatListRef: any;

  constructor(props: ButtonGroupProps) {
    super(props);
    const { defaultActiveIndex } = this.props;
    this.state = { activeIndex: defaultActiveIndex || 0 };
  }

  static getDerivedStateFromProps(nextProps: any, prevState: any) {
    const { defaultIndexChange } = nextProps;
    const { activeIndex } = prevState;
    if (defaultIndexChange && nextProps.defaultActiveIndex !== activeIndex) {
      return {
        activeIndex: nextProps.defaultActiveIndex,
      };
    }
    return {
      activeIndex,
    };
  }

  componentDidUpdate() {
    const { activeIndex } = this.state;
    if (this.flatListRef && activeIndex) {
      this.flatListRef.flatListRef.scrollToIndex({
        animated: true,
        index: activeIndex - 1 > 0 ? activeIndex - 1 : 0,
      });
    }
  }

  onPress = (index: number) => {
    this.setState({ activeIndex: index });
    const { onItemPress } = this.props;
    if (onItemPress) {
      onItemPress(index);
    }
  };

  getIndex = () => {
    const { activeIndex } = this.state;
    return activeIndex || 0;
  };

  renderItem = ({ item, index }: { item: string; index: number }) => {
    const {
      flatListProps,
      titleList,
      marginRight: marginRightProp,
      marginHorizontal,
      ...otherButtonProps
    } = this.props;
    const { activeIndex } = this.state;
    const list = titleList;
    const marginRight = marginRightProp || 10;
    if (list) {
      return (
        <Button
          outline
          {...otherButtonProps}
          title={titleList ? item : undefined}
          marginRight={
            index === list.length - 1 ? (marginHorizontal || 0) : marginRight
          }
          marginLeft={
            index === 0 ? marginHorizontal : 0
          }
          active={activeIndex === index}
          onPress={() => this.onPress(index)}
        />
      );
    }
    return <QuickView />;
  };

  render() {
    const {
      titleList, flatListProps, defaultActiveIndex, defaultIndexChange,
    } = this.props;
    const { activeIndex } = this.state;
    const FlatListButton: any = FlatList;
    return (
      <FlatListButton
        ref={(ref: any) => {
          this.flatListRef = ref;
        }}
        horizontal
        renderItem={this.renderItem}
        {...flatListProps}
        data={titleList}
        initialScrollIndex={defaultActiveIndex}
        onScrollToIndexFailed={() => {
          this.flatListRef.flatListRef.scrollToIndex({
            animated: true,
            index: 0,
          });
        }}
        showsHorizontalScrollIndicator={flatListProps?.horizontal || false}
        extraData={defaultIndexChange ? defaultActiveIndex : activeIndex}
      />
    );
  }
}

export default ButtonGroup;
