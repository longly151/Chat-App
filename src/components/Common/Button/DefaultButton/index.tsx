/* eslint-disable max-len */
import React from 'react';
import { StyleSheet } from 'react-native';
import {
  Button as ElementButton,
  ButtonProps as EButtonProps,
  withTheme,
  ThemeProps,
} from 'react-native-elements';
import _ from 'lodash';
import Color from '@themes/Color';
import AppView from '@utils/appView';
import { Translation } from 'react-i18next';
import Font from '@themes/Font';

export interface ButtonProps extends Omit<EButtonProps, 'raised' | 'type'> {
  width?: number | string;
  height?: number | string;
  margin?: number;
  marginTop?: number;
  marginBottom?: number;
  marginLeft?: number;
  marginRight?: number;
  marginHorizontal?: number;
  marginVertical?: number;
  titlePaddingVertical?: number;
  titlePaddingHorizontal?: number;
  primary?: boolean;
  secondary?: boolean;
  success?: boolean;
  warning?: boolean;
  error?: boolean;
  title?: string;
  titleCenter?: boolean;
  titleColor?: string;
  backgroundColor?: string;
  borderRadius?: number;
  borderColor?: string;
  activeBorderColor?: string;
  borderWidth?: number;
  center?: boolean;
  bold?: boolean;
  fontSize?: number;
  active?: boolean;
  activeBackgroundColor?: string;
  activeTitleColor?: string;
  outline?: boolean;
  clear?: boolean;
  sharp?: boolean;
  rounded?: boolean;
  circle?: boolean;
  shadow?: boolean;
  color?: string;
  reverseColor?: boolean;
  t?: string | Array<any>;
  theme?: any;
}

class Button extends React.PureComponent<ButtonProps> {
  static defaultProps = {
    margin: 0,
    active: false,
    rounded: true,
    titleCenter: true,
  };

  render() {
    const {
      width: widthProp,
      height: heightProp,
      margin,
      marginTop,
      marginBottom,
      marginLeft,
      marginRight,
      marginHorizontal: marginHorizontalProp,
      marginVertical: marginVerticalProp,
      titlePaddingVertical: titlePaddingVerticalProp,
      titlePaddingHorizontal: titlePaddingHorizontalProp,
      borderRadius: borderRadiusProp,
      primary,
      secondary,
      success,
      error,
      warning,
      borderColor: borderColorProp,
      activeBorderColor: activeBorderColorProp,
      borderWidth,
      center,
      bold,
      fontSize,
      active,
      icon: iconProp,
      iconRight,
      backgroundColor: backgroundColorProp,
      title,
      titleCenter,
      titleColor: titleColorProp,
      activeBackgroundColor: activeBackgroundColorProp,
      activeTitleColor: activeTitleColorProp,
      containerStyle: containerStyleProp,
      buttonStyle: buttonStyleProp,
      titleStyle: titleStyleProp,
      iconContainerStyle: iconContainerStyleProp,
      outline,
      clear,
      sharp: sharpProp,
      rounded: roundedProp,
      circle: circleProp,
      shadow,
      color,
      reverseColor,
      t,
      theme,
      ...otherProps
    } = this.props;

    const { colors } = theme;

    /**
     * containerStyle
     */
    const marginVertical = marginVerticalProp || theme.Button.marginVertical;
    const marginHorizontal = marginHorizontalProp || theme.Button.marginHorizontal;
    let width = widthProp;
    let height = heightProp || theme.Button.height;

    let borderRadius: any = 0;
    let sharp = sharpProp;
    let rounded = roundedProp;
    let circle = circleProp;
    if (borderRadiusProp) {
      borderRadius = borderRadiusProp;
    } else {
      if (circle) {
        rounded = false;
        sharp = false;
        const minDimension = _.min([width, height]) || theme.Button.height || 50;
        width = minDimension;
        height = minDimension;
        borderRadius = minDimension;
      }
      if (rounded) {
        borderRadius = theme.Button.roundedBorderRadius;
      }
      if (sharp) {
        rounded = false;
        circle = false;
        borderRadius = 0;
      }
    }

    const containerStyle: any = StyleSheet.flatten([
      {
        margin,
        marginTop,
        marginBottom,
        marginLeft,
        marginRight,
        marginHorizontal,
        marginVertical,
        borderRadius,
      },
      shadow && { paddingBottom: 3 },
      center && { alignSelf: 'center' },
      containerStyleProp,
    ]);

    /**
     * buttonStyle
     */
    const titlePaddingHorizontal = titlePaddingHorizontalProp || theme.Button.titlePaddingHorizontal;
    const titlePaddingVertical = titlePaddingVerticalProp || theme.Button.titlePaddingVertical;

    const activeBackgroundColor = activeBackgroundColorProp
    || color
    || theme.Button.activeBackgroundColor;
    let backgroundColor = backgroundColorProp || color || theme.Button.backgroundColor;

    if (primary) backgroundColor = colors.primary;
    if (secondary) backgroundColor = colors.secondary;
    if (success) backgroundColor = colors.success;
    if (warning) backgroundColor = colors.warning;
    if (error) backgroundColor = colors.error;

    const activeBorderColor = activeBorderColorProp || color || theme.Button.activeBorderColor;
    let borderColor = borderColorProp || color || theme.Button.borderColor;
    if (primary) borderColor = colors.primary;
    if (secondary) borderColor = colors.secondary;
    if (success) borderColor = colors.success;
    if (warning) borderColor = colors.warning;
    if (error) borderColor = colors.error;

    let buttonStyle: any = StyleSheet.flatten([
      {
        width,
        height,
        borderRadius,
        borderWidth: borderWidth || theme.Button.borderWidth,
        borderColor: active ? activeBorderColor : borderColor,
        backgroundColor: active ? activeBackgroundColor : backgroundColor,
      },
      !_.isUndefined(titlePaddingHorizontal) && { paddingHorizontal: titlePaddingHorizontal },
      !_.isUndefined(titlePaddingVertical) && { paddingVertical: titlePaddingVertical },
      shadow && AppView.shadow,
      buttonStyleProp,
      !titleCenter && { justifyContent: 'flex-start' },
    ]);
    /**
     * titleStyle
     */

    const activeTitleColor = activeTitleColorProp
    || (color ? Color.white : theme.Button.activeTitle);
    const titleColor = titleColorProp || (color ? Color.white : theme.Button.titleColor);

    let titleStyle: any = StyleSheet.flatten([
      {
        fontFamily: Font.fontFamily.Roboto,
        fontWeight: (active || bold) ? 'bold' : 'normal',
        fontSize: fontSize || theme.Button.titleFontSize,
        color: active ? activeTitleColor : titleColor,
        marginRight: (iconProp && !iconRight && !circle) ? 10 : 0,
        marginLeft: (iconProp && iconRight && !circle) ? 10 : 0,
      },
      titleStyleProp,
    ]);

    /**
     * icon, iconContainerStyle
     */
    const iconContainerStyle: any = StyleSheet.flatten([
      (iconProp && !iconRight && title) && { marginRight: 5 },
      iconContainerStyleProp,
    ]);
    const defaultIcon = {
      size: 20,
      color: active ? activeTitleColor : titleColor,
      type: 'material-community',
    };
    const icon = iconProp ? _.merge(defaultIcon, iconProp) : undefined;
    /**
     * type
     */
    if (clear) {
      buttonStyle = StyleSheet.flatten([
        buttonStyle,
        {
          backgroundColor: 'transparent',
          borderWidth: 0,
        },
      ]);

      const activeOutlineTitleColor = activeTitleColorProp
      || color
      || theme.Button.activeOutlineTitleColor;
      let outlineTitleColor = titleColorProp || color || theme.Button.outlineTitleColor;
      if (primary) outlineTitleColor = colors.primary;
      if (secondary) outlineTitleColor = colors.secondary;
      if (success) outlineTitleColor = colors.success;
      if (warning) outlineTitleColor = colors.warning;
      if (error) outlineTitleColor = colors.error;

      titleStyle = StyleSheet.flatten([
        titleStyle,
        {
          color: active ? activeOutlineTitleColor : outlineTitleColor,
        },
      ]);
    } else if (outline) {
      const activeOutlineBorderColor = activeBorderColorProp
      || color
      || theme.Button.activeOutlineBorderColor;
      let outlineBorderColor = borderColorProp || color || theme.Button.outlineBorderColor;
      if (primary) outlineBorderColor = colors.primary;
      if (secondary) outlineBorderColor = colors.secondary;
      if (success) outlineBorderColor = colors.success;
      if (warning) outlineBorderColor = colors.warning;
      if (error) outlineBorderColor = colors.error;

      const activeOutlineTitleColor = activeTitleColorProp
      || color
      || theme.Button.activeOutlineTitleColor;
      let outlineTitleColor = titleColorProp || color || theme.Button.outlineTitleColor;
      if (primary) outlineTitleColor = colors.primary;
      if (secondary) outlineTitleColor = colors.secondary;
      if (success) outlineTitleColor = colors.success;
      if (warning) outlineTitleColor = colors.warning;
      if (error) outlineTitleColor = colors.error;

      buttonStyle = StyleSheet.flatten([
        buttonStyle,
        {
          backgroundColor: active ? activeBackgroundColor : 'transparent',
          borderWidth: borderWidth || theme.Button.outlineBorderWidth,
          borderColor: active ? activeOutlineBorderColor : outlineBorderColor,
        },
      ]);

      titleStyle = StyleSheet.flatten([
        titleStyle,
        {
          color: active ? activeOutlineTitleColor : outlineTitleColor,
        },
      ]);
    }

    /**
     * reverseColor
     */
    if (reverseColor) {
      const reverseBackgroundColor = titleStyle.color;
      const reverseTitleColor = buttonStyle.backgroundColor;
      buttonStyle.backgroundColor = reverseBackgroundColor;
      titleStyle.color = reverseTitleColor;
    }

    if (!title && t) {
      return (
        <Translation>
          {(trans) => (
            <ElementButton
              {...otherProps}
              icon={icon}
              iconRight={iconRight}
              containerStyle={containerStyle}
              buttonStyle={buttonStyle}
              title={typeof t === 'string' ? trans(t) : trans(t[0], t[1])}
              titleStyle={titleStyle}
              iconContainerStyle={iconContainerStyle}
            />
          )}
        </Translation>
      );
    }

    return (
      <ElementButton
        {...otherProps}
        icon={icon}
        iconRight={iconRight}
        containerStyle={containerStyle}
        buttonStyle={buttonStyle}
        title={title}
        titleStyle={titleStyle}
        iconContainerStyle={iconContainerStyle}
      />
    );
  }
}

export default withTheme(Button as React.ComponentType<ButtonProps & ThemeProps<any>>);
