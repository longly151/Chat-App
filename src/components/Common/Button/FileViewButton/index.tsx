import React, { Component } from 'react';
import {
  TouchableOpacity, Platform
} from 'react-native';
import _ from 'lodash';
import moment from 'moment';
import RNFetchBlob from 'rn-fetch-blob';
import { IFile } from '@utils/appHelper';
import AppView from '@utils/appView';
import { withTheme, ThemeProps } from 'react-native-elements';
import Text from '../../Text';
import QuickView from '../../View/QuickView';
import Loading from '../../Loading';
import Icon from '../../Icon';

export interface FileViewButtonProps {
  data: IFile;
  color?: string;
  backgroundColor?: string;
  theme?: any;
  horizontal?: boolean;
}
interface State {
  loading: boolean
}
class FileViewButton extends Component<FileViewButtonProps, State> {
  private tempFileUri: string = '';

  constructor(props:FileViewButtonProps) {
    super(props);
    this.state = { loading: false };
  }

  componentWillUnmount() {
    if (this.tempFileUri) {
      RNFetchBlob.fs.unlink(this.tempFileUri).then(() => {
      });
    }
  }

  renderIcon = (size: number = 30) => {
    const { color } = this.props;
    const { loading } = this.state;
    if (loading) {
      return (
        <Loading />
      );
    }
    const { data } = this.props;
    let name = 'file';
    // Image
    if (_.startsWith(data.mime, 'image')) {
      name = 'image';
    }
    // Pdf
    if (data.mime === 'application/pdf') {
      name = 'file-pdf';
    }
    // Doc
    if (data.mime === 'application/msword' || data.mime === 'application/vnd.openxmlformats-officedocument.wordprocessingml.document' || data.mime === 'application/vnd.openxmlformats-officedocument.wordprocessingml.template' || data.mime === 'application/vnd.ms-word.document.macroEnabled.12' || data.mime === 'application/vnd.ms-word.template.macroEnabled.12') {
      name = 'file-word';
    }
    // Excel
    if (data.mime === 'application/vnd.ms-excel' || data.mime === 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' || data.mime === 'application/vnd.openxmlformats-officedocument.spreadsheetml.template' || data.mime === 'application/vnd.ms-excel.sheet.macroEnabled.12' || data.mime === 'application/vnd.ms-excel.template.macroEnabled.12' || data.mime === 'application/vnd.ms-excel.addin.macroEnabled.12' || data.mime === 'application/vnd.ms-excel.sheet.binary.macroEnabled.12') {
      name = 'file-excel';
    }
    // PowerPoint
    if (data.mime === 'application/vnd.ms-powerpoint' || data.mime === 'application/vnd.openxmlformats-officedocument.presentationml.presentation' || data.mime === 'application/vnd.openxmlformats-officedocument.presentationml.template' || data.mime === 'application/vnd.openxmlformats-officedocument.presentationml.slideshow' || data.mime === 'application/vnd.ms-powerpoint.addin.macroEnabled.12' || data.mime === 'application/vnd.ms-powerpoint.presentation.macroEnabled.12' || data.mime === 'application/vnd.ms-powerpoint.template.macroEnabled.12' || data.mime === 'application/vnd.ms-powerpoint.slideshow.macroEnabled.12') {
      name = 'file-powerpoint';
    }

    // Color
    return <Icon name={name} type="material-community" size={size} color={color} />;
  };

  onPress = async () => {
    await this.setState({ loading: true });
    const { data } = this.props;
    if (data.remoteUrl) {
      const fileExtension = data.remoteUrl.split('.').pop();
      const { dirs } = RNFetchBlob.fs;
      RNFetchBlob
        .config({
          fileCache: true,
          // by adding this option, the temp files will have a file extension
          appendExt: fileExtension,
          path: `${dirs.DocumentDir}/${data.name || 'Untitled'}.${fileExtension}`,
        })
        .fetch('GET', data.remoteUrl, {
          // some headers ..
        })
        .then((res) => {
          this.setState({ loading: false });
          this.tempFileUri = res.path();
          // console.log('res.path()', res.path());
          // the temp file path with file extension `png`
          if (Platform.OS === 'ios') RNFetchBlob.ios.openDocument(res.data);
          else { RNFetchBlob.android.actionViewIntent(`${res.path()}`, data.mime); }
          this.setState({ loading: false });
        });
    }
  };

  render() {
    const {
      data,
      color,
      backgroundColor: backgroundColorProp,
      horizontal,
      theme,
    } = this.props;
    const backgroundColor = backgroundColorProp || theme.colors.secondaryBackground;

    if (horizontal) {
      return (
        <TouchableOpacity onPress={this.onPress} style={{ marginVertical: 10 }}>
          <QuickView
            borderRadius={AppView.roundedBorderRadius}
            width={140}
            height={140}
            backgroundColor={backgroundColor}
            padding={15}
            justifyContent="space-around"
          >
            <QuickView center>
              {this.renderIcon(50)}
            </QuickView>
            <Text center fontSize={16} color={color} bold numberOfLines={2}>{data.name}</Text>
          </QuickView>
        </TouchableOpacity>
      );
    }

    return (
      <TouchableOpacity onPress={this.onPress} style={{ width: '100%', marginVertical: 10 }}>
        <QuickView borderRadius={AppView.roundedBorderRadius} backgroundColor={backgroundColor} style={[{ width: '100%' }]}>
          <QuickView padding={18} row>
            <QuickView row center>
              <QuickView flex={1}>
                {this.renderIcon()}
              </QuickView>
              <QuickView flex={9} marginLeft={10}>
                <Text fontSize={16} color={color} bold numberOfLines={2}>{data.name}</Text>
                <Text fontSize={12} color={color} marginTop={5} t={['component:file_view_button:updated_at', { updatedAt: moment(data.updatedAt).format('DD/MM/YYYY') }]} />
              </QuickView>
            </QuickView>
          </QuickView>
        </QuickView>
      </TouchableOpacity>
    );
  }
}

export default withTheme(
  FileViewButton as any as React.ComponentType<FileViewButtonProps & ThemeProps<any>>
);
