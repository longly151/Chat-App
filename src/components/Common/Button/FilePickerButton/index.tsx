import React, { Component } from 'react';
import DocumentPicker, { DocumentPickerOptions, DocumentPickerResponse } from 'react-native-document-picker';
import { GestureResponderEvent, FlatList } from 'react-native';
import AppHelper, { IFile } from '@utils/appHelper';
import _ from 'lodash';
import QuickView from '@components/Common/View/QuickView';
import Button, { ButtonProps } from '../DefaultButton';
import FileViewButton from '../FileViewButton';

interface FilePickerButtonProps extends ButtonProps {
  filePickerProps: DocumentPickerOptions<any>;
  multiple?: boolean;
  folderPrefix?: string;
  uploadCallback?: (file: IFile[]) => Promise<any> | any;
  pickSuccess?: (file: DocumentPickerResponse[]) => Promise<any> | any;
  handleException?: (e: any) => any;
}

interface State {
  loading: boolean;
  data: IFile[];
  folderPrefix?: string;
}

class FilePickerButton extends Component<FilePickerButtonProps, State> {
  static defaultProps = {
    filePickerProps: {
      type: [
        DocumentPicker.types.doc,
        DocumentPicker.types.docx,
        DocumentPicker.types.pdf,
        DocumentPicker.types.csv,
        DocumentPicker.types.xls,
        DocumentPicker.types.xlsx,
        DocumentPicker.types.zip,
        DocumentPicker.types.audio,
        DocumentPicker.types.video,
      ]
    },
    folderPrefix: 'files',
  };

  constructor(props: FilePickerButtonProps) {
    super(props);

    this.state = {
      loading: false,
      data: []
      // data: [
      //   {
      //     name: 'img0005-2020103108093228.JPG',
      //     uri: 'https://airtravel.s3.us-east-2.amazonaws.com/images/img0005-2020103108093228.JPG',
      //     type: 'image/jpeg',
      //     updatedAt: new Date()
      //   },
      //   {
      //     name: 'social-media.pdf',
      //     uri: 'https://airtravel.s3.us-east-2.amazonaws.com/files/social-media-2020110101592791.pdf',
      //     type: 'application/pdf',
      //     updatedAt: new Date()
      //   },
      //   {
      //     name: 'bai-1-a-a.docx',
      //     uri: 'https://airtravel.s3.us-east-2.amazonaws.com/files/bai-1-a-a-2020110102254107.docx',
      //     type: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
      //     updatedAt: new Date()
      //   }
      // ]
    };
  }

  uploadFiles = async (files: DocumentPickerResponse[]) => {
    const { folderPrefix, uploadCallback } = this.props;
    this.setState({ loading: true });

    // Get uploadUrls
    const uploadUrlBody: any = [];
    files.forEach((item: DocumentPickerResponse) => {
      uploadUrlBody.push({
        type: item.type,
        fileName: item.name,
        folderPrefix,
      });
    });
    const uploadUrls = await AppHelper.getUploadUrls(uploadUrlBody);

    await Promise.all(files.map(async (file: DocumentPickerResponse, index: number) => {
      const data = {
        name: file.name,
        type: file.type,
        uri: file.fileCopyUri // or file.uri,
      };

      /**
       * uploadToS3
       */
      await setTimeout(() => {}, 100);
      try {
        await AppHelper.uploadToS3(uploadUrls[index].presignedUrl, data);
        this.setState((previousState: any) => ({
          data: [...previousState.data, {
            name: data.name,
            mime: data.type,
            size: file.size,
            path: data.uri,
            sourceUrl: file.uri,
            remoteUrl: uploadUrls[index].returnUrl,
            updatedAt: new Date()
          }]
        }));
      } catch (error) {
        this.handleException(error);
      }
    }));
    this.setState({ loading: false });
    const { data } = this.state;
    if (uploadCallback) await uploadCallback(data);
  };

  pickSuccess = async (res: DocumentPickerResponse[]) => {
    // Custom Action...
    const { pickSuccess } = this.props;
    if (pickSuccess) pickSuccess(res);
    const { pickSuccess: pickSuccessProp } = this.props;
    if (pickSuccessProp) pickSuccessProp(res);

    await this.uploadFiles(res);
  };

  handleException = (e: any) => {
    // Custom Action...
    const { handleException } = this.props;
    if (handleException) handleException(e);
  };

  pickFile = async (event: GestureResponderEvent) => {
    const { filePickerProps, multiple, onPress } = this.props;
    const { type } = filePickerProps;

    // Pick a single file
    if (!multiple) {
      try {
        const res = await DocumentPicker.pick({ type });
        this.pickSuccess([res]);
      } catch (err) {
        this.handleException(err);
      }
    } else {
      // Pick multiple files
      try {
        const results = await DocumentPicker.pickMultiple({ type });
        this.pickSuccess(results);
      } catch (err) {
        this.handleException(err);
      }
    }

    if (onPress) onPress(event);
  };

  getData = () => {
    const { data } = this.state;
    return data;
  };

  removeFileItem = (file: IFile) => {
    const { data } = this.state;
    this.setState({
      data: data.filter((item: IFile) => item.remoteUrl !== file.remoteUrl)
    });
  };

  renderFileItem = ({ item }: {item: IFile}) => (
    <QuickView marginHorizontal={5}>
      <FileViewButton
        data={item}
        horizontal
      />
      <Button
        icon={{ name: 'close', size: 20 }}
        width={30}
        titlePaddingHorizontal={0}
        titlePaddingVertical={0}
        circle
        backgroundColor="#E6E9F0"
        containerStyle={{ position: 'absolute', right: 10, top: 10 }}
        onPress={() => this.removeFileItem(item)}
      />
    </QuickView>
  );

  render() {
    const { ...otherProps } = this.props;
    const { data, loading } = this.state;
    if (_.isEmpty(data)) {
      return (
        <Button
          {...otherProps}
          onPress={this.pickFile}
          loading={loading}
        />
      );
    }

    return (
      <FlatList
        data={data}
        renderItem={this.renderFileItem}
        horizontal
        contentContainerStyle={{ marginHorizontal: -5 }}
        showsHorizontalScrollIndicator={false}
        keyExtractor={(item, index) => `${index}`}
      />
    );
  }
}

export default FilePickerButton;
