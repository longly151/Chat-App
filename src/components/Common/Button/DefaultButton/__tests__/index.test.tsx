/* eslint-disable no-undef */
import React from 'react';
import TestHelper, { EnumComponent } from '@utils/testHelper';
import { lightTheme } from '@themes/Theme';
import { Platform } from 'react-native';
import Color from '@themes/Color';
import AppView from '@utils/appView';
import Button from '..';

const target = <Button />;
const type = EnumComponent.HOC;

describe('Button Component', () => {
  TestHelper.basicTest(target, type);

  it('should apply all style-related props', () => {
    const props = {
      width: 100,
      height: 100,
      margin: 10,
      marginVertical: 10,
      marginHorizontal: 10,
      marginTop: 10,
      marginBottom: 10,
      marginLeft: 10,
      marginRight: 10,
      padding: 10,
      paddingVertical: 10,
      paddingHorizontal: 10,
      paddingTop: 10,
      paddingBottom: 10,
      paddingLeft: 10,
      paddingRight: 10,
    };
    TestHelper.testStyleProps(target, props, type, 'containerStyle');

    const props1 = {
      borderRadius: 20,
      borderColor: Color.orange,
      borderWidth: 5,
    };
    TestHelper.testStyleProps(target, props1, type, 'buttonStyle');
  });

  it('should apply all custom props', () => {
    /**
     * center
     */
    const testStyle = TestHelper.getStyle(target, { center: true }, type, 'containerStyle');
    expect(testStyle.alignSelf).toBe('center');

    /**
     * bold, titleColor
     */
    const testStyle1 = TestHelper.getStyle(target, { bold: true }, type, 'titleStyle');
    expect(testStyle1.fontWeight).toBe('bold');
    const testStyle2 = TestHelper.getStyle(target, { titleColor: Color.orange }, type, 'titleStyle');
    expect(testStyle2.color).toBe(Color.orange);

    /**
     * type = solid | null, primary, secondary, success, warning, error, backgroundColor,
     */
    const testStyle3 = TestHelper.getStyle(target, { primary: true }, type, 'buttonStyle');
    expect(testStyle3.backgroundColor).toBe(lightTheme.colors.primary);
    const testStyle4 = TestHelper.getStyle(target, { secondary: true }, type, 'buttonStyle');
    expect(testStyle4.backgroundColor).toBe(lightTheme.colors.secondary);
    const testStyle5 = TestHelper.getStyle(target, { success: true }, type, 'buttonStyle');
    expect(testStyle5.backgroundColor).toBe(lightTheme.colors.success);
    const testStyle6 = TestHelper.getStyle(target, { error: true }, type, 'buttonStyle');
    expect(testStyle6.backgroundColor).toBe(lightTheme.colors.error);
    const testStyle7 = TestHelper.getStyle(target, { warning: true }, type, 'buttonStyle');
    expect(testStyle7.backgroundColor).toBe(lightTheme.colors.warning);
    const testStyle8 = TestHelper.getStyle(target, { backgroundColor: Color.orange }, type, 'buttonStyle');
    expect(testStyle8.backgroundColor).toBe(Color.orange);

    /**
     * type = outline, primary, secondary, success, warning, error,
     */
    const testStyle15 = TestHelper.getStyle(target, { outline: true }, type, 'buttonStyle');
    expect(testStyle15.backgroundColor).toBe('transparent');
    expect(testStyle15.borderColor).toBe(lightTheme.Button.outlineBorderColor);

    const testStyleOutline3 = TestHelper.getStyle(target, { outline: true, primary: true }, type, 'buttonStyle');
    expect(testStyleOutline3.borderColor).toBe(lightTheme.colors.primary);
    const testStyleOutline4 = TestHelper.getStyle(target, { outline: true, secondary: true }, type, 'buttonStyle');
    expect(testStyleOutline4.borderColor).toBe(lightTheme.colors.secondary);
    const testStyleOutline5 = TestHelper.getStyle(target, { outline: true, success: true }, type, 'buttonStyle');
    expect(testStyleOutline5.borderColor).toBe(lightTheme.colors.success);
    const testStyleOutline6 = TestHelper.getStyle(target, { outline: true, error: true }, type, 'buttonStyle');
    expect(testStyleOutline6.borderColor).toBe(lightTheme.colors.error);
    const testStyleOutline7 = TestHelper.getStyle(target, { outline: true, warning: true }, type, 'buttonStyle');
    expect(testStyleOutline7.borderColor).toBe(lightTheme.colors.warning);
    jest.resetModules();

    /**
     * type = clear, primary, secondary, success, warning, error,
     */
    const testStyle16 = TestHelper.getStyle(target, { clear: true }, type, 'buttonStyle');
    expect(testStyle16.backgroundColor).toBe('transparent');
    expect(testStyle16.borderWidth).toBe(0);
    const testStyle17 = TestHelper.getStyle(target, { clear: true }, type, 'titleStyle');
    expect(testStyle17.color).toBe(lightTheme.Button.outlineTitleColor);

    Platform.OS = 'android';
    const testStyleClear3 = TestHelper.getStyle(target, { clear: true, primary: true }, type, 'titleStyle');
    expect(testStyleClear3.color).toBe(lightTheme.colors.primary);
    const testStyleClear4 = TestHelper.getStyle(target, { clear: true, secondary: true }, type, 'titleStyle');
    expect(testStyleClear4.color).toBe(lightTheme.colors.secondary);
    const testStyleClear5 = TestHelper.getStyle(target, { clear: true, success: true }, type, 'titleStyle');
    expect(testStyleClear5.color).toBe(lightTheme.colors.success);
    const testStyleClear6 = TestHelper.getStyle(target, { clear: true, error: true }, type, 'titleStyle');
    expect(testStyleClear6.color).toBe(lightTheme.colors.error);
    const testStyleClear7 = TestHelper.getStyle(target, { clear: true, warning: true }, type, 'titleStyle');
    expect(testStyleClear7.color).toBe(lightTheme.colors.warning);
    jest.resetModules();

    /**
     * active, activeBorderColor, activeBackgroundColor, activeTitleColor
     */
    const testStyle18 = TestHelper.getStyle(target, {
      active: 'true', activeBorderColor: Color.orange, activeBackgroundColor: Color.orange,
    }, type, 'buttonStyle');
    expect(testStyle18.borderColor).toBe(Color.orange);
    expect(testStyle18.backgroundColor).toBe(Color.orange);

    const testStyle19 = TestHelper.getStyle(target, {
      active: 'true', activeTitleColor: Color.orange,
    }, type, 'titleStyle');
    expect(testStyle19.color).toBe(Color.orange);

    /**
     * sharp, rounded, circle
     */
    const testStyle20 = TestHelper.getStyle(target, { sharp: 'true' }, type, 'buttonStyle');
    expect(testStyle20.borderRadius).toBe(0);

    const testStyle21 = TestHelper.getStyle(target, { rounded: 'true' }, type, 'buttonStyle');
    expect(testStyle21.borderRadius).toBe(lightTheme.Button.roundedBorderRadius);

    const testStyle22 = TestHelper.getStyle(target, { circle: 'true' }, type, 'buttonStyle');
    expect(testStyle22.width).toBe(testStyle22.height);
    expect(testStyle22.width).toBe(testStyle22.borderRadius);

    /**
     * shadow
     */
    const testStyle23 = TestHelper.getStyle(target, { shadow: true }, type, 'buttonStyle');
    expect(testStyle23).toMatchObject(AppView.shadow);
    const testStyle24 = TestHelper.getStyle(target, { shadow: true }, type, 'containerStyle');
    expect(testStyle24.paddingBottom).toBeGreaterThanOrEqual(3);
  });
});
