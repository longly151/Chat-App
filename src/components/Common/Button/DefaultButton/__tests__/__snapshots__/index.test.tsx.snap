// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should render without issues 1`] = `
<ForwardRef(Themed.Button)
  buttonStyle={
    Object {
      "backgroundColor": "#bdc6cf",
      "borderColor": "#ffffff",
      "borderRadius": 8,
      "borderWidth": 0,
      "height": 40,
      "paddingHorizontal": 20,
      "paddingVertical": 0,
      "width": undefined,
    }
  }
  containerStyle={
    Object {
      "borderRadius": 8,
      "margin": 0,
      "marginBottom": undefined,
      "marginHorizontal": 0,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "marginVertical": 5,
    }
  }
  iconContainerStyle={Object {}}
  titleStyle={
    Object {
      "color": "#121212",
      "fontFamily": "Roboto",
      "fontSize": 16,
      "fontWeight": "normal",
      "marginLeft": 0,
      "marginRight": 0,
    }
  }
/>
`;
